src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{BTreeMap, HashMap};
+use std::fs::{self};
+use std::io::Cursor;
+use std::path::{Path, PathBuf};
+use std::time::Duration;
+
+use anyhow::{bail, Context};
+use byteorder::{NetworkEndian, ReadBytesExt};
+use lazy_static::lazy_static;
+use log::error;
+use regex::Regex;
+use tokio::select;
+use tokio::sync::mpsc;
+
+use dataflow_types::Update;
+use expr::GlobalId;
+use repr::{Row, Timestamp};
+
+use crate::wal::{encode_progress, encode_update};
+
+/// Organize and maintain a compacted representation of persisted data.
+///
+/// The `Compacter` task keeps track of the persisted updates for each persisted relation and
+/// periodically compacts that representation to use space proportional to the number of
+/// distinct rows in the relation. In order to do so, the `Compacter` maintains a `Trace` for
+/// each persisted relation.
```

Comment:
A question I had around here (once having read it) is whether the compacter is meant to be hosted in materialize, run elsewhere, or "doesn't matter". It may be a red herring, but trace management in DD is carefully metered out on some important threads, but if this is meant to be spun up wherever (some cloud function maybe) that's different and maybe has different expectations. Idk. Maybe not important but wanted to call out thoughts.

Reply:
added documentation to clarify that the compacter can be a separate process or on aws even on a separate box eventually as long as it shares access to the same storage the wal is written to